app/backend/src/clj/teet/notification/ion.clj

```diff
@@ -6,25 +6,29 @@
             [datomic.client.api :as d]
             [teet.notification.notification-db :as notification-db]))
 
+(defn notify-tx-data
+  "Transaction data for notification"
+  [db application-id]
+  (let [ project-id (cooperation-db/application-project-id db application-id)
+         activity-id (cooperation-db/application-activity-id db application-id)
+         user-to (activity-db/activity-manager db activity-id)]
+     (notification-db/system-notification-tx db
+       {:to user-to
+        :project project-id
+        :target application-id
+        :type cooperation-notifications/application-response-expired-soon})) )
+
 (defn notify
   "Given number of days before Application expiration
   new notification is sent to active Activity PM and
   consultant who entered the application into system
   providing the Activity is not finished"
   ([days]
-   (let [conn (environment/datomic-connection)
-         db (d/db conn)
-         applications-to-expire (cooperation-db/applications-to-be-expired db days)]
-     (map #(let [application-id (first %)
-                 project-id (cooperation-db/application-project-id db application-id)
-                 activity-id (cooperation-db/application-activity-id db application-id)
-                 user-to (activity-db/activity-manager db activity-id)]
-             (notification-db/system-notification-tx db
-               {:to user-to
-                :project project-id
-                :target application-id
-                :type cooperation-notifications/application-response-expired-soon}))
-       applications-to-expire)))
+   (let [conn (environment/datomic-connection) db (d/db conn)]
+     (map #(let [app-id (first %)
+                 current-tx-data ['(teet.notification.ion/notify-tx-data app-id)]]
+             (d/transact conn {:tx-data current-tx-data}))
```

Comment:
Don't call transact in a map. Gather all info into one transaction data and transact it once.

Reply:
And no need to call notify-tx-data as a transaction function, it isn't doing any constraint validation that needs to be atomic.

Instead consider:
```(t/transact conn {:tx-data (map (partial notify-tx-data) (cooperation-db/applications-to-be-expired db days))})```